### PR TITLE
Implement contains() member function

### PR DESCRIPTION
**Pull request by nickaein**
This pull request is intended to tackle #1471.

This adds `contains()` method which similar to [`std::map::contains()`](https://en.cppreference.com/w/cpp/container/map/contains) available in C++20. While there could be various APIs and choices for such function, this PR tries to be aligned with STL counterpart. 

Having such method, the user can check the existence of a key in a JSON object more concisely:

```
if(j.contains("foo"))
{
    // "foo" key exists
}
else
{
    // "foo" key not exists	
}
```

`std::map::contains()` has been proposed in [P0458R2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0458r2.html) and is merged to the [latest C++20 draft N4800](https://github.com/cplusplus/draft/blob/master/papers/n4800.pdf) (as now). Boost library [has already adopted it](https://www.boost.org/doc/libs/master/doc/html/container/release_notes.html#container.release_notes.release_notes_boost_1_69_00) since then.

* * *

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header file `single_include/nlohmann/json.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

## Please don't

- The C++11 support varies between different **compilers** and versions. Please note the [list of supported compilers](https://github.com/nlohmann/json/blob/master/README.md#supported-compilers). Some compilers like GCC 4.7 (and earlier), Clang 3.3 (and earlier), or Microsoft Visual Studio 13.0 and earlier are known not to work due to missing or incomplete C++11 support. Please refrain from proposing changes that work around these compiler's limitations with `#ifdef`s or other means.
- Specifically, I am aware of compilation problems with **Microsoft Visual Studio** (there even is an [issue label](https://github.com/nlohmann/json/issues?utf8=✓&q=label%3A%22visual+studio%22+) for these kind of bugs). I understand that even in 2016, complete C++11 support isn't there yet. But please also understand that I do not want to drop features or uglify the code just to make Microsoft's sub-standard compiler happy. The past has shown that there are ways to express the functionality such that the code compiles with the most recent MSVC - unfortunately, this is not the main objective of the project.
- Please refrain from proposing changes that would **break [JSON](http://json.org) conformance**. If you propose a conformant extension of JSON to be supported by the library, please motivate this extension.
- Please do not open pull requests that address **multiple issues**.

### CODE REVIEW

File: include/nlohmann/json.hpp [hunk 1]
```diff
@@ -3957,6 +3957,39 @@ class basic_json
...
+    @complexity Logarithmic in the size of the JSON object.
+
+    @since version 3.6.0
+    */
+    template<typename KeyT>
+    bool contains(KeyT&& key) const
+    {
+        if (is_object())
+        {
+            return (m_value.object->find(key) != m_value.object->end());
```
Comment (nlohmann) [inline include/nlohmann/json.hpp]: I'm not sure, but shouldn't this be `find(std::forward(key))`?

**Comment by nlohmann** [comment]
Very nice - I only have a small question, see above.

File: include/nlohmann/json.hpp (hunk 1 above)
Comment (nickaein) [reply include/nlohmann/json.hpp]: Good point! Done.

**Comment by nlohmann** [comment]
Thanks!

**Comment by nickaein** [comment]
I initially assumed this will be available for the next minor release (`3.6.0`) since it's an backward-compatible API change. So I put `@since version 3.6.0` in the documentation. Sorry that I forgot to mention it.

Will this addition be available in `3.5.1` and should I open another PR to fix the documentation?

File: include/nlohmann/json.hpp [hunk 2]
```diff
@@ -3957,6 +3957,39 @@ class basic_json
...
+    template<typename KeyT>
+    bool contains(KeyT&& key) const
+    {
+        if (is_object())
+        {
+            return (m_value.object->find(std::forward<KeyT>(key)) != m_value.object->end());
+        }
+        else
+        {
+            return false;
```
Comment (jaredgrubb) [inline include/nlohmann/json.hpp]: Should this throw?

**Comment by jaredgrubb** [comment]
If I ask if a number contains a key, I think it should throw.
Even more so, if I ask an *array* if it contains a value, getting false back feels odd.

File: include/nlohmann/json.hpp (hunk 2 above)
Comment (nickaein) [reply include/nlohmann/json.hpp]: It seems [std::map::find cannot be marked as noexcept](https://stackoverflow.com/a/34639172/161640). I am not sure if we can guarantee that conditions not happening in our case.

**Comment by nickaein** [comment]
What would a better behavior is such cases is debatable. I tried to keep it on par with the already existing [`json::find()`](https://github.com/nickaein/nlohmann-json/blob/develop/include/nlohmann/json.hpp#L3903). For non-object types, it returns `std::end()` which is the same return value if the type is object and the key is not found.

Essentially `json.contains(x)` is equivalent to `json.find(x) != end()` and I believe this is probably expected behavior. It might be confusing that `json.find(x)` (and `json.count(x)`, etc) doesn't throw for non-objects but `json::contains(x)` does.

**Comment by jaredgrubb** [comment]
Those are good arguments and, with that justification, I have to reluctantly agree. :)

In another world I would try to argue that they all should throw, but the cure is probably worse than the cause at this point.

**Comment by nickaein** [comment]
@nlohmann What's your take on https://github.com/nlohmann/json/pull/1474#issuecomment-463180891

**Comment by nlohmann** [comment]
I already adjusted the target version to 3.6.0 if this is what you mean.

**Comment by nickaein** [comment]
Yes Thanks. The timeline still says

> [nlohmann](/nlohmann) added this to the [Release 3.5.1](/nlohmann/json/milestone/28) milestone [9 days ago](#event-2136514712)

but I can now see the sidebar has the correct milestone.

**Comment by nlohmann** [comment]
I renamed the milestone which apparently does not change past log messages. :)

**Pull request merged**